commit 1aedfdde796cf06a1909f15f770260bfd55b2224
Author: Steve Muckle <smuckle@google.com>
Date:   Mon Nov 23 14:33:37 2020 -0800

    eas: trace_parse: accept variable number of spaces after PID, CPU
    
    Recent kernels have variable spacing after the PID and CPU numbers.
    Amend the trace parsing to accept that.
    
    Bug: 172295168
    Signed-off-by: Steve Muckle <smuckle@google.com>
    Change-Id: Ia43b6451e41e8c1467d54fd93a433c9a6a96af7a

diff --git a/testcases/kernel/sched/eas/trace_parse.c b/testcases/kernel/sched/eas/trace_parse.c
index a4c158a1b..0aeb930a2 100644
--- a/testcases/kernel/sched/eas/trace_parse.c
+++ b/testcases/kernel/sched/eas/trace_parse.c
@@ -231,15 +231,15 @@ static void *parse_event_data(unsigned int event_type, char *data)
  * First 16 chars are the currently running thread name. Drop leading spaces.
  * Next char is a dash
  * Next 5 chars are PID. Drop trailing spaces.
- * Next char is a space.
+ * Next is some number of spaces.
  * Next five chars are the CPU, i.e. [001].
- * Next char is a space.
+ * Next is some number of spaces.
  * Next letter is whether IRQs are off.
  * Next letter is if NEED_RESCHED is set.
  * Next letter is if this is in hard or soft IRQ context.
  * Next letter is the preempt disable depth.
- * Next char is a space.
- * Next twelve letters are the timestamp. Drop leading spaces.
+ * Next is some number of spaces.
+ * Next twelve letters are the timestamp.
  * Next char is colon.
  * Next char is space.
  * Next twelve letters are the event name.
@@ -247,7 +247,7 @@ static void *parse_event_data(unsigned int event_type, char *data)
  * Rest of line is string specific to event.
  */
 static int parse_trace_record(struct trace_record *tr, char *line) {
-	unsigned int idx = 0;
+	unsigned int field_start, idx = 0;
 	char *found;
 
 	/* Skip leading spaces in the task name. */
@@ -266,83 +266,100 @@ static int parse_trace_record(struct trace_record *tr, char *line) {
 		return -1;
 	}
 	idx = 17;
-	if (line[22] != ' ') {
-		printf("Malformatted trace record, no space between"
-		       "pid and CPU:\n");
+	while(line[idx] && line[idx] != '[') idx++;
+	if (!line[idx]) {
+		printf("Malformatted trace record, no open bracket for"
+		       "CPU after pid:\n");
 		printf("%s", line);
 		return -1;
 	}
-	line[22] = 0;
+	line[idx-1] = 0;
 	if (sscanf(&line[17], "%hd", &tr->pid) != 1) {
 		printf("Malformatted trace record, error parsing"
 		       "pid:\n");
 		printf("%s", line);
 		return -1;
 	}
-	if (line[28] != ' ') {
-		printf("Malformatted trace record, no space between"
-		       "CPU and flags:\n");
+	field_start = idx;
+	while(line[idx] && line[idx] != ']') idx++;
+	if (!line[idx]) {
+		printf("Malformatted trace record, no closing bracket for"
+		       "CPU:\n");
+		printf("%s", line);
+		return -1;
+	}
+	idx++;
+	if (line[idx] != ' ') {
+		printf("Malformatted trace record, no space following CPU:\n");
 		printf("%s", line);
 		return -1;
 	}
-	line[28] = 0;
-	if (sscanf(&line[23], "[%hd]", &tr->cpu) != 1) {
+	line[idx] = 0;
+	idx++;
+	if (sscanf(&line[field_start], "[%hd]", &tr->cpu) != 1) {
 		printf("Malformatted trace record, error parsing CPU:\n");
 		printf("%s", line);
 		return -1;
 	}
-	if (line[29] == 'd') {
+
+	if (line[idx] == 'd') {
 		tr->flags |= TRACE_RECORD_IRQS_OFF;
-	} else if (line[29] != '.') {
+	} else if (line[idx] != '.') {
 		printf("Malformatted trace record, error parsing irqs-off:\n");
 		printf("%s", line);
 		return -1;
 	}
-	if (line[30] == 'N') {
+	idx++;
+	if (line[idx] == 'N') {
 		tr->flags |= TRACE_RECORD_TIF_NEED_RESCHED;
 		tr->flags |= TRACE_RECORD_PREEMPT_NEED_RESCHED;
-	} else if (line[30] == 'n') {
+	} else if (line[idx] == 'n') {
 		tr->flags |= TRACE_RECORD_TIF_NEED_RESCHED;
-	} else if (line[30] == 'p') {
+	} else if (line[idx] == 'p') {
 		tr->flags |= TRACE_RECORD_PREEMPT_NEED_RESCHED;
-	} else if (line[30] != '.') {
+	} else if (line[idx] != '.') {
 		printf("Malformatted trace record, error parsing "
 		       "need-resched:\n");
 		printf("%s", line);
 		return -1;
 	}
+	idx++;
 
-	if (line[31] != IRQ_CONTEXT_NORMAL && line[31] != IRQ_CONTEXT_SOFT &&
-	    line[31] != IRQ_CONTEXT_HARD &&
-	    line[31] != IRQ_CONTEXT_HARD_IN_SOFT &&
-	    line[31] != IRQ_CONTEXT_NMI && line[31] != IRQ_CONTEXT_NMI_IN_HARD) {
+	if (line[idx] != IRQ_CONTEXT_NORMAL && line[idx] != IRQ_CONTEXT_SOFT &&
+	    line[idx] != IRQ_CONTEXT_HARD &&
+	    line[idx] != IRQ_CONTEXT_HARD_IN_SOFT &&
+	    line[idx] != IRQ_CONTEXT_NMI && line[idx] != IRQ_CONTEXT_NMI_IN_HARD) {
 		printf("Malformatted trace record, error parsing irq "
 		       "context:\n");
 		printf("%s", line);
 		return -1;
 	}
-	tr->irq_context = line[31];
+	tr->irq_context = line[idx];
+	idx++;
 
-	if (line[33] != ' ') {
+	if (line[idx+1] != ' ') {
 		printf("Malformatted trace record, no space between"
 		       "flags and timestamp:\n");
 		printf("%s", line);
 		return -1;
 	}
-	line[33] = 0;
-	if (line[32] == '.') {
+	line[idx+1] = 0;
+	if (line[idx] == '.') {
 		tr->preempt_depth = 0;
-	} else if (sscanf(&line[32], "%hx", &tr->preempt_depth) != 1) {
+	} else if (sscanf(&line[idx], "%hx", &tr->preempt_depth) != 1) {
 		printf("Malformatted trace record, error parsing "
 		       "preempt-depth:\n");
 		printf("%s", line);
 		return -1;
 	}
+	idx += 2;
 
-	/* timestamp starts as early as line[34], skip leading spaces */
-	idx = 34;
-	while (idx < 38 && line[idx] == ' ')
-		idx++;
+	while (line[idx] && line[idx] == ' ') idx++;
+	if (!line[idx]) {
+		printf("Malformatted trace record, missing timestamp:\n");
+		printf("%s", line);
+		return -1;
+	}
 	if (sscanf(&line[idx], "%d.%d: ", &tr->ts.sec,
 		   &tr->ts.usec) != 2) {
 		printf("Malformatted trace record, error parsing "
